module/zfs/vdev_raidz.c

```diff
@@ -1525,41 +1538,76 @@ vdev_raidz_io_start_write(zio_t *zio, raidz_row_t *rr, uint64_t ashift)
 {
 	vdev_t *vd = zio->io_vd;
 	raidz_map_t *rm = zio->io_vsd;
-	int c, i;
+	int c, skipped = 0, skip_first_cols = 0;
 
 	vdev_raidz_generate_parity_row(rm, rr);
 
-	for (int c = 0; c < rr->rr_cols; c++) {
+	if (rr->rr_scols < (rm->rm_skipstart + rm->rm_nskip)) {
+		skip_first_cols =
+		    (rm->rm_skipstart + rm->rm_nskip) % rr->rr_scols;
+	}
+
+	for (c = 0; c < rr->rr_scols; c++) {
+		abd_t *abd = NULL;
+		enum zio_flag flags = 0;
 		raidz_col_t *rc = &rr->rr_col[c];
-		if (rc->rc_size == 0)
-			continue;
+		vdev_t *cvd = vd->vdev_child[rc->rc_devidx];
 
 		/* Verify physical to logical translation */
 		vdev_raidz_io_verify(vd, rr, c);
 
-		zio_nowait(zio_vdev_child_io(zio, NULL,
-		    vd->vdev_child[rc->rc_devidx], rc->rc_offset,
-		    rc->rc_abd, rc->rc_size, zio->io_type, zio->io_priority,
-		    0, vdev_raidz_child_done, rc));
-	}
+		/*
+		 * Generate I/O for skip sectors to improve aggregation
+		 * contiguity. We will use gang ABD's to reduce contention
+		 * on the children VDEV queue locks (vq_lock) by issuing
+		 * a single I/O that contains the data and skip sectors.
+		 */
```

Comment:
The logic here looks correct to me.  However, I think the code would be more understandable if we took the same approach used in the dRAID code.  Specifically, instead of trying to figure out in `vdev_raidz_io_start_write()` if we need to setup a gang IO we determine this when creating the mapping in `vdev_raidz_map_alloc()`. This way we create the right ABD for the column in a single location.  This also means we can then issue  the zio in a simple loop without additional complicated logic.

Take a look at the `vdev_draid_map_alloc()->vdev_draid_map_alloc_write()` and `vdev_draid_io_start_write()` call paths for reference.

Reply:
@behlendorf I went ahead and updated this to match `vdev_draid_map_alloc_write()`. I think it makes the code much cleaner to read this way as well. I was able to find another bug in the original code where I was not accounting for skip columns starting at the first parity column. For now, I kept this as two separate commits, but I figured if all CI tests pass it could be squashed into a single commit. The code seem pretty stable from some of the ZTS tests I was running on my local machine, but the full ZTS test runs will verify everything is in order with this update.